test/test_foreach.py

```diff
@@ -806,6 +835,115 @@ def test_add_list_slow_path(self, device, dtype):
+                [getattr(t, op.ref.__name__ + '_')() for t in tensors]
+        else:
+            self.assertEqual(expected, tensors)
+
+    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=True))
+    def test_binary_op_tensors_on_different_devices(self, device, dtype):
+        if self.device_type != 'cuda':
+            self.skipTest('CUDA is necessary for tests with tensors on different devices')
+        for foreach_op, foreach_op_, native_op in self.bin_ops:
+            # `tensors1`: ['cuda', 'cpu']
+            # `tensors2`: ['cuda', 'cpu']
+            _cuda_tensors = self._get_test_data(device, dtype, 2)
+            _cpu_tensors = self._get_test_data('cpu', dtype, 2)
+            tensors1, tensors2 = [[_cuda_tensors[i], _cpu_tensors[i]] for i in range(2)]
+
+            if dtype == torch.bool and native_op == torch.sub:
+                with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                    foreach_op(tensors1, tensors2)
+                with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                    foreach_op_(tensors1, tensors2)
+                continue
+            try:
+                actual = foreach_op(tensors1, tensors2)
+            except RuntimeError as e:
+                with self.assertRaisesRegex(type(e), str(e)):
+                    [native_op(t1, t2) for t1, t2 in zip(tensors1, tensors2)]
+            else:
+                expected = [native_op(t1, t2) for t1, t2 in zip(tensors1, tensors2)]
+                self.assertEqual(expected, actual)
+            try:
+                foreach_op_(tensors1, tensors2)
+            except RuntimeError as e:
+                with self.assertRaisesRegex(type(e), str(e)):
+                    [getattr(t1, native_op.__name__ + '_')(t2) for t1, t2 in zip(tensors1, tensors2)]
+            else:
+                self.assertEqual(actual, tensors1)
+
+    @skipMeta
+    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=True))
+    def test_pointwise_op_tensors_on_different_devices(self, device, dtype):
+        if self.device_type != 'cuda':
+            self.skipTest('CUDA is necessary for tests with tensors on different devices')
+
+        pointwise_ops = [
+            (torch._foreach_addcmul, torch._foreach_addcmul_, torch.addcmul),
+            (torch._foreach_addcdiv, torch._foreach_addcdiv_, torch.addcdiv),
+        ]
+        for foreach_op, foreach_op_, native_op in pointwise_ops:
+            # tensors1: ['cuda', 'cpu]
+            # tensors2: ['cuda', 'cpu]
+            # tensors3: ['cuda', 'cpu]
+            _cuda_tensors = self._get_test_data(device, dtype, 3)
+            _cpu_tensors = self._get_test_data('cpu', dtype, 3)
+            tensors1, tensors2, tensors3 = [[_cuda_tensors[i], _cpu_tensors[i]] for i in range(3)]
+
+            if native_op == torch.addcdiv:
+                if dtype in torch.testing.get_all_int_dtypes() + [torch.bool]:
```

Comment:
can you please leave a comment here why this special case is necessary?

Reply:
Okay, I think the reason is similar to that of `sub and bool` case: the message has one or more special characters